cpp/src/io/parquet/page_hdr.cu

```diff
@@ -121,70 +122,150 @@ __device__ void skip_struct_field(byte_stream_s *bs, int t)
+    if (t != ST_FLD_I32)
+      return true;
+    else
+      return false;
+  }
+};
+
+template <typename Operator>
+struct ParquetFieldStruct {
+  int field;
+  Operator op;
+
+  __device__ ParquetFieldStruct(int f) : field(f) {}
+
+  inline __device__ bool operator()(byte_stream_s *bs, int t)
+  {
+    if ((t != ST_FLD_STRUCT) || !op(bs))
+      return true;
+    else
+      return false;
+  }
+};
+
+template <int index>
+struct FunctionSwitchImpl {
+  template <typename... Operator>
+  static inline __device__ bool run(byte_stream_s *bs,
+                                    int t,
+                                    const int &fld,
+                                    thrust::tuple<Operator...> &ops)
+  {
+    if (fld == thrust::get<index>(ops).field) {
+      return thrust::get<index>(ops)(bs, t);
+    } else {
+      return FunctionSwitchImpl<index - 1>::run(bs, t, fld, ops);
     }
+  }
+};
 
-PARQUET_BEGIN_STRUCT(gpuParseDataPageHeader)
-PARQUET_FLD_INT32(1, page.num_input_values)
-PARQUET_FLD_ENUM(2, page.encoding, Encoding);
-PARQUET_FLD_ENUM(3, page.definition_level_encoding, Encoding);
-PARQUET_FLD_ENUM(4, page.repetition_level_encoding, Encoding);
-PARQUET_END_STRUCT()
-
-PARQUET_BEGIN_STRUCT(gpuParseDictionaryPageHeader)
-PARQUET_FLD_INT32(1, page.num_input_values)
-PARQUET_FLD_ENUM(2, page.encoding, Encoding);
-PARQUET_END_STRUCT()
-
-PARQUET_BEGIN_STRUCT(gpuParseDataPageHeaderV2)
-PARQUET_FLD_INT32(1, page.num_input_values)
-PARQUET_FLD_INT32(3, page.num_rows)
-PARQUET_FLD_ENUM(4, page.encoding, Encoding);
-PARQUET_FLD_ENUM(5, page.definition_level_encoding, Encoding);
-PARQUET_FLD_ENUM(6, page.repetition_level_encoding, Encoding);
-PARQUET_END_STRUCT()
-
-PARQUET_BEGIN_STRUCT(gpuParsePageHeader)
-PARQUET_FLD_ENUM(1, page_type, PageType)
-PARQUET_FLD_INT32(2, page.uncompressed_page_size)
-PARQUET_FLD_INT32(3, page.compressed_page_size)
-PARQUET_FLD_STRUCT(5, gpuParseDataPageHeader)
-PARQUET_FLD_STRUCT(7, gpuParseDictionaryPageHeader)
-PARQUET_FLD_STRUCT(8, gpuParseDataPageHeaderV2)
-PARQUET_END_STRUCT()
+template <>
+struct FunctionSwitchImpl<0> {
+  template <typename... Operator>
+  static inline __device__ bool run(byte_stream_s *bs,
+                                    int t,
+                                    const int &fld,
+                                    thrust::tuple<Operator...> &ops)
+  {
+    if (fld == thrust::get<0>(ops).field) {
+      return thrust::get<0>(ops)(bs, t);
+    } else {
+      skip_struct_field(bs, t);
+      return false;
+    }
+  }
+};
+
+template <typename... Operator>
+inline __device__ bool function_builder(thrust::tuple<Operator...> &op, byte_stream_s *bs)
+{
+  constexpr int index = thrust::tuple_size<thrust::tuple<Operator...>>::value - 1;
+  int fld             = 0;
+  for (;;) {
+    int c, t, f;
+    c = getb(bs);
+    if (!c) break;
+    f                  = c >> 4;
+    t                  = c & 0xf;
+    fld                = (f) ? fld + f : get_i32(bs);
```

Comment:
This also goes for multiple code blocks:

- Declare variables at the point of initialization;
- Mark variables as const;
- Improve naming;

My attempt for the code here:
```suggestion
    auto const next_byte = getb(bs);
    if (!next_byte) break;
    auto const upper_nibble = next_byte >> 4; // still a bad name, because I don't know the meaning of the value
    auto const lower_nibble = next_byte & 0xf; // still a bad name, because I don't know the meaning of the value
    field          = (upper_nibble != 0) ? field + upper_nibble : get_i32(bs);
```